### PR TITLE
⚠️ Raise `Exception` when trying to generate 0 tokens ⚠️

### PR DESCRIPTION
**Pull request by danielkorat**
# What does this PR do?

Currently, setting `max_new_tokens=0` produces 1 token instead of 0, and only a warning is produced.
To prevent unexpected patterns of generation, this warning should be changed to an `Exception`.

### Example:

```python
from transformers import AutoModelForCausalLM, AutoTokenizer
checkpoint = "bigcode/tiny_starcoder_py"
tokenizer = AutoTokenizer.from_pretrained("bigcode/tiny_starcoder_py")
model = AutoModelForCausalLM.from_pretrained("bigcode/tiny_starcoder_py")
inputs = tokenizer("def print_hello_world():", return_tensors="pt")
outputs = model.generate(**inputs,
                        pad_token_id=tokenizer.eos_token_id,
                        max_new_tokens=0)
print(f"Input length: {len(inputs['input_ids'][0])}")
print(f"Output length: {len(outputs[0])}")
```

### Output before fix:
```bash
/home/sdp/fix-zero-max-new-tokens/transformers/src/transformers/generation/utils.py:1136: UserWarning: Input length of input_ids is 7, but `max_length` is set to 7. This can lead to unexpected behavior. You should consider increasing `max_new_tokens`.
  warnings.warn(
Input length: 7
Output length: 8
```

### Output after fix:
```bash
Traceback (most recent call last):
  File "/home/sdp/fix-zero-max-new-tokens/test.py", line 8, in <module>
    outputs = model.generate(**inputs,
  File "/storage/sdp/anaconda3/envs/fix-zero-max-new-tokens/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/sdp/fix-zero-max-new-tokens/transformers/src/transformers/generation/utils.py", line 1396, in generate
    self._validate_generated_length(generation_config, input_ids_length, has_default_max_length)
  File "/home/sdp/fix-zero-max-new-tokens/transformers/src/transformers/generation/utils.py", line 1136, in _validate_generated_length
    raise ValueError(
ValueError: Input length of input_ids is 7, but `max_length` is set to 7. This can lead to unexpected behavior. You should consider increasing `max_new_tokens`.
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@gante

### CODE REVIEW

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -1133,11 +1133,10 @@ def _validate_generated_length(self, generation_config, input_ids_length, has_de
             )
         if input_ids_length >= generation_config.max_length:
             input_ids_string = "decoder_input_ids" if self.config.is_encoder_decoder else "input_ids"
-            warnings.warn(
+            raise ValueError(
                 f"Input length of {input_ids_string} is {input_ids_length}, but `max_length` is set to"
                 f" {generation_config.max_length}. This can lead to unexpected behavior. You should consider"
-                " increasing `max_new_tokens`.",
-                UserWarning,
+                " increasing `max_new_tokens`."
```
Comment (gante) [inline src/transformers/generation/utils.py:1139]:
```suggestion
                " increasing `max_length` or, better yet, setting `max_new_tokens`."
```

**Review by gante — COMMENTED** [review]
In addition to the changes in the PR, I would also check `max_new_tokens` in `GenerationConfig.validate()` -- it must be > 0 when it is set, otherwise an exception should be raised

**Comment by gante** [comment]
@danielkorat the test is failing because it is poorly parameterized -- it raises the newly added exception. I'd like to ask you to try to fix the test parameterization (ping me if you get stuck :) )

**Comment by danielkorat** [comment]
> @danielkorat the test is failing because it is poorly parameterized -- it raises the newly added exception. I'd like to ask you to try to fix the test parameterization (ping me if you get stuck :) )

Hi @gante @amyeroberts, 
I made the requested changes and fixed the test above as well. 
This test is now failing and it seems to be unrelated to my changes:
```bash
_ FlaxWav2Vec2ModelTest.test_equivalence_pt_to_flax [FlaxWav2Vec2ForPreTraining] _
....
E   AssertionError: 1.1444092e-05 not less than or equal to 1e-05 : outputs.codevector_perplexity: Difference between PyTorch and Flax is 1.1444091796875e-05 (>= 1e-05).
```

**Comment by danielkorat** [comment]
@gante

**Comment by ArthurZucker** [comment]
It was probably a flaky test 😉

**Review by ArthurZucker — APPROVED** [review]
LGTM let's make sur the title of the pr has ⚠️ as it's breaking

**Comment by danielkorat** [comment]
Thanks @ArthurZucker, I changed the title

**Comment by danielkorat** [comment]
@ArthurZucker can you merge please?

**Comment by ArthurZucker** [comment]
Thanks for the PR 😉

**Comment by gante** [comment]
@danielkorat thank you for the fix 💛

**Pull request merged**